### ISSUE TITLE
Gradle "eats" stdin, spelling trouble for automation as well as manual use in a shell

### ISSUE THREAD

**Issue by christianhujer** [gradle/gradle#14961]
### Expected Behavior
Expected behavior: As Gradle does not need the information on stdin, Gradle should not read or close stdin and leave it untouched.

### Current Behavior
Current behavior: Gradle eats stdin.

### Context
The context is scripting as well as manual use in a shell.

#### How it affects manual use
In a shell, run `sleep 10`. While the sleep command is "running", type `echo foo`. The sleep command does not "eat" stdin. Thus, after the sleep command completes, `echo foo` is received by the shell and executed next. This is the normal behavior of all programs that do not need stdin (except Gradle).

This affects me quite annoyingly on a daily basis. For example, I run `./gradlew build` and then remember that I wanted to also see the updates, so, while the build is running, I type `./gradlew checkUpdates`. With every normal tool, this would work. But with Gradle, the `./gradlew checkUpdates` will not be executed because the shell will not receive the command once Gradle is done, caused by Gradle eating stdin. I have to _wait_ for Gradle, and that is the anti-thesis of automation.

#### How it affects scripting
In case Gradle is to be part of a command sequence that overall is to act as a filter, Gradle destroys the filter-chain by eating up all stdin.

### Steps to Reproduce
Run `echo foo | (gradle;cat) | grep foo`
Expected: The output contains `foo`.
Actual: The output does not contain `foo`.

More examples:
Run `echo foo | (sleep 1;cat) | grep foo` ⇒ output contains `foo`
Run `echo foo | (mvn;cat) | grep foo` ⇒ output contains `foo`
Run `echo foo | (make;cat) | grep foo` ⇒ output contains `foo`
Run `echo foo | (cmake;cat) | grep foo` ⇒ output contains `foo`
Run `echo foo | (ant;cat) | grep foo` ⇒ output contains `foo`
Run `echo foo | (gradle;cat) | grep foo` ⇒ output does *not* contain `foo`

Because Gradle isn't supposed to read anything from stdin, it is expected that `foo` is found in the output.
This may lead to very unexpected behavior when using Gradle as a component in bigger scripts.

This behavior of Gradle is inconsistent with the expected behavior of command-line programs in general, as well as among build tools in particular.

### Workaround
I know of a number of workarounds. They are based on the idea of giving Gradle its own empty stdin. That prevents Gradle from eating the stdin of its context.
1. Don't call Gradle directly. Instead, call it from a `Makefile`, and use `echo -n | gradle` or `echo -n | ./gradlew` as the actual Gradle command.
2. Use an alias: `alias gradle='echo -n | gradle'` or `alias gr=echo -n | ./gradlew`

Demonstration that the workaround works:
Run `echo foo | (echo -n | gradle;cat) | grep foo` ⇒ output contains `foo`

### Your Environment
This happens with all Gradle versions I've ever used, with or without wrapper.
Currently, I'm using Gradle 6.7.

**Comment by danielcranford** [comment]
This also means gradle cannot be run in the background, as any background process in the shell attempting to read from stdin gets paused via a SIGTTIN
```
$ ./gradlew build &
[1] 22636
$ jobs
[1]+  Stopped                 ./gradlew build
$
```

**Comment by kota65535** [comment]
I'm also surprised at this non-intuitive behavior.
I wrote a bash script with a while loop that reads a file line by line and run gradle commands, but it breaks immediately after the first loop.

```bash
while read -r line; do 
  ./gradlew 
done < some.txt
```

**Comment by decimal1969** [comment]
Is this hard to fix?
It has been a while since I looked at UNIX/Linux processes and stdin but I'm guessing either stdin is getting closed or flushed and should not be.

I use the same workaround above where I append < /dev/null to my gradle command.
For me this is useful if I redirect the gradle output to a file and want to look at it with a pager command after.

./gradlew --console=plain {target} > gradlew.out 2>&1 < /dev/null
less gradlew.out

If I cut/paste those two lines from somewhere into bash/ksh/zsh it should "just work" but because of this the "less gradlew.out" command is never run. This is what makes me think gradle is flushing stdin. I don't think closing stdin would cause it but it has been a while since I looked at this type of behavior.
